File: src/main/java/de/tectoast/emolga/bot/EmolgaListener.java
package de.tectoast.emolga.bot;

import de.tectoast.emolga.buttons.ButtonListener;
import de.tectoast.emolga.commands.Command;
import de.tectoast.emolga.commands.PrivateCommand;
import de.tectoast.emolga.commands.PrivateCommands;
import de.tectoast.emolga.selectmenus.MenuListener;
import de.tectoast.emolga.utils.Constants;
import de.tectoast.emolga.utils.DexQuiz;
import de.tectoast.emolga.utils.draft.Draft;
import de.tectoast.emolga.utils.music.GuildMusicManager;
import de.tectoast.emolga.utils.sql.DBManagers;
import de.tectoast.jsolf.JSONObject;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.ReadyEvent;
import net.dv8tion.jda.api.events.channel.ChannelCreateEvent;
import net.dv8tion.jda.api.events.guild.GuildJoinEvent;
import net.dv8tion.jda.api.events.guild.invite.GuildInviteCreateEvent;
import net.dv8tion.jda.api.events.guild.member.GuildMemberJoinEvent;
import net.dv8tion.jda.api.events.guild.voice.GuildVoiceLeaveEvent;
import net.dv8tion.jda.api.events.guild.voice.GuildVoiceMoveEvent;
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent;
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent;
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
import net.dv8tion.jda.api.events.interaction.component.SelectMenuInteractionEvent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.events.role.RoleCreateEvent;
import net.dv8tion.jda.api.events.user.update.UserUpdateNameEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.interactions.AutoCompleteQuery;
import net.dv8tion.jda.api.interactions.components.buttons.Button;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;
import java.io.IOException;
import java.net.URL;
import java.sql.Timestamp;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static de.tectoast.emolga.commands.Command.*;
import static de.tectoast.emolga.utils.Constants.EMOLGA_KI;
import static de.tectoast.emolga.utils.Constants.MYSERVER;

public class EmolgaListener extends ListenerAdapter {
    public static final List<String> allowsCaps = Arrays.asList("712612442622001162", "752230819644440708", "732545253344804914");
    public static final String WELCOMEMESSAGE = """
            Hallo **{USERNAME}** und vielen Dank, dass du mich auf deinen Server **{SERVERNAME}** geholt hast!
            Vermutlich möchtest du für deinen Server hauptsächlich, dass die Ergebnisse von Showdown Replays in einen Channel geschickt werden.
            Um mich zu konfigurieren gibt es folgende Möglichkeiten:

            **1. Die Ergebnisse sollen in den gleichen Channel geschickt werden:**
            Einfach `!replaychannel` in den jeweiligen Channel schreiben

            **2. Die Ergebnisse sollen in einen anderen Channel geschickt werden:**
            `!replaychannel #Ergebnischannel` in den Channel schicken, wo später die Replays reingeschickt werden sollen (Der #Ergebnischannel ist logischerweise der Channel, wo später die Ergebnisse reingeschickt werden sollen)

            Falls die Ergebnisse in ||Spoilertags|| geschickt werden sollen, schick irgendwo auf dem Server den Command `!spoilertags` rein. Dies gilt dann serverweit.

            Ich habe übrigens noch viele weitere Funktionen! Wenn du mich pingst, zeige ich dir eine Übersicht aller Commands :)
            Falls du weitere Fragen oder Probleme hast, schreibe %s eine PN :)""".formatted(Constants.MYTAG);
    private static final Logger logger = LoggerFactory.getLogger(EmolgaListener.class);

    @Override
    public void onButtonInteraction(@NotNull ButtonInteractionEvent e) {
        ButtonListener.check(e);
    }

    @Override
    public void onModalInteraction(@NotNull ModalInteractionEvent e) {
        e.reply(e.getValues().stream().map(m -> "%s: %s (%s)".formatted(m.getId(), m.getAsString(), m.getType().toString())).collect(Collectors.joining("\n"))).queue();
    }

    @Override
    public void onSelectMenuInteraction(@NotNull SelectMenuInteractionEvent e) {
        MenuListener.check(e);
    }

    @Override
    public void onGuildMemberJoin(@Nonnull GuildMemberJoinEvent e) {
        Guild g = e.getGuild();
        Member mem = e.getMember();
        long gid = g.getIdLong();
        if (gid == Constants.BSID) {
            g.addRoleToMember(mem, g.getRoleById(715242519528603708L)).queue();
        }
        if (gid == Constants.ASLID) {
            g.getTextChannelById(615605820381593610L).sendMessage(
                    "Willkommen auf der ASL, " + mem.getAsMention() + ". <:hi:540969951608045578>\n" +
                    "Dies ist ein Pokémon Server mit dem Fokus auf einem kompetetiven Draftligasystem. " +
                    "Mach dich mit dem <#635765395038666762> vertraut und beachte die vorgegebenen Themen der Kanäle. Viel Spaß! <:yay:540970044297838597>"
            ).queue();
        }
    }

    @Override
    public void onGuildVoiceMove(@NotNull GuildVoiceMoveEvent e) {
        if (e.getChannelLeft().getMembers().size() == 1 && e.getGuild().getAudioManager().isConnected()) {
            e.getGuild().getAudioManager().closeAudioConnection();
        }
    }

    @Override
    public void onGuildJoin(@Nonnull GuildJoinEvent e) {
        Guild g = e.getGuild();
        g.retrieveOwner().flatMap(m -> m.getUser().openPrivateChannel()).queue(pc -> pc.sendMessage(WELCOMEMESSAGE.replace("{USERNAME}", g.getOwner().getUser().getName()).replace("{SERVERNAME}", g.getName())).queue());
        e.getJDA().retrieveUserById(Constants.FLOID).flatMap(User::openPrivateChannel).flatMap(
                        u -> u.sendMessage("%s (%s)".formatted(g.getName(), g.getId())).setActionRow(Button.primary("guildinvite;" + g.getId(), "Invite").withEmoji(Emoji.fromUnicode("✉️"))))
                .queue();
    }

    @Override
    public void onRoleCreate(@Nonnull RoleCreateEvent e) {
        if (!e.getGuild().getId().equals("736555250118295622") && !e.getGuild().getId().equals("447357526997073930") && !e.getGuild().getId().equals("518008523653775366"))
            return;
        e.getRole().getManager().revokePermissions(Permission.CREATE_INSTANT_INVITE).queue();
    }

    @Override
    public void onMessageReceived(@NotNull MessageReceivedEvent e) {
        if (e.getChannelType().isGuild()) {
            try {
                if (e.isWebhookMessage()) return;
                Message m = e.getMessage();
                String msg = m.getContentDisplay();
                TextChannel tco = e.getTextChannel();
                Member member = e.getMember();
                Guild g = e.getGuild();
                long gid = g.getIdLong();
                check(e);
                if (gid == MYSERVER) {
                    PrivateCommands.execute(e.getMessage());
                    if (tco.getParentCategoryIdLong() == EMOLGA_KI) {
                        String[] split = tco.getName().split("-");
                        e.getJDA().getTextChannelById(split[split.length - 1]).sendMessage(m.getContentRaw()).queue();
                    }
                }
                if (tco.getIdLong() == 929841771276554260L) {
                    g.addRoleToMember(member, g.getRoleById(934810601216147477L)).queue();
                }
                String raw = m.getContentRaw();
                long id = e.getJDA().getSelfUser().getIdLong();
                if (raw.equals("<@!" + id + ">") || raw.equals("<@" + id + ">") && !e.getAuthor().isBot() && isChannelAllowed(tco)) {
                    help(tco, member);
                }
                if (emoteSteal.contains(tco.getIdLong())) {
                    List<Emote> l = m.getMentions().getEmotes();
                    for (Emote emote : l) {
                        try {
                            g.createEmote(emote.getName(), Icon.from(new URL(emote.getImageUrl()).openStream())).queue();
                        } catch (IOException ioException) {
                            ioException.printStackTrace();
                        }
                    }
                }
                if (tco.getIdLong() == 778380440078647296L || tco.getIdLong() == 919641011632881695L) {
                    String[] split = msg.split(" ");
                    JSONObject counter = shinycountjson.getJSONObject("counter");
                    Optional<String> mop = counter.keySet().stream().filter(s -> s.toLowerCase().startsWith(split[1].toLowerCase())).findFirst();
                    if (mop.isPresent()) {
                        JSONObject o = shinycountjson.getJSONObject("counter").getJSONObject(mop.get());
                        boolean isCmd = true;
                        String mid = member.getId().equals("893773494578470922") ? "598199247124299776" : member.getId();
                        if (msg.contains("!set ")) {
                            o.put(mid, Integer.parseInt(split[2]));
                        } else if (msg.contains("!reset ")) {
                            o.put(mid, 0);
                        } else if (msg.contains("!add ")) {
                            o.put(mid, o.optInt(mid, 0) + Integer.parseInt(split[2]));
                        } else isCmd = false;
                        if (isCmd) {
                            m.delete().queue();
                            updateShinyCounts(tco.getIdLong());
                        }
                    }
                }
                if (!e.getAuthor().isBot() && !msg.startsWith("!dexquiz")) {
                    DexQuiz quiz = DexQuiz.getByTC(tco);
                    if (quiz != null && quiz.nonBlocking()) {
                        if (quiz.check(msg)) {
                            quiz.block();
                            tco.sendMessage(member.getAsMention() + " hat das Pokemon erraten! Es war **" + quiz.getCurrentGerName() + "**! (Der Eintrag stammt aus **Pokemon " + quiz.getCurrentEdition() + "**)").queue();
                            quiz.givePoint(member.getIdLong());
                            quiz.nextRound();
                        }
                    }
                }
                JSONObject json = getEmolgaJSON();
                if (replayAnalysis.containsKey(tco.getIdLong()) && !e.getAuthor().getId().equals(e.getJDA().getSelfUser().getId()) && !msg.contains("!analyse ") && !msg.contains("!sets ")) {
                    TextChannel t = tco.getGuild().getTextChannelById(replayAnalysis.get(tco.getIdLong()));
                    //t.sendTyping().queue();
                    if (msg.contains("https://") || msg.contains("http://")) {
                        Optional<String> urlop = Arrays.stream(msg.split("\n")).filter(s -> s.contains("https://replay.pokemonshowdown.com") || s.contains("http://florixserver.selfhost.eu:228/")).map(s -> s.substring(s.indexOf("http"), s.indexOf(" ", s.indexOf("http") + 1) == -1 ? s.length() : s.indexOf(" ", s.indexOf("http") + 1))).findFirst();
                        if (urlop.isPresent()) {
                            String url = urlop.get();
                            logger.info(url);
                            analyseReplay(url, null, t, m, null);
                        }
                    }
                }
            } catch (IllegalStateException ex) {
                sendToMe(e.getChannelType().name() + " Illegal Argument Exception Channel");
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        } else if (e.isFromType(ChannelType.PRIVATE)) {
            if (!e.getAuthor().isBot() && e.getAuthor().getIdLong() != Constants.FLOID)
                e.getJDA().getTextChannelById(828044461379682314L).sendMessage(e.getAuthor().getAsMention() + ": " + e.getMessage().getContentDisplay()).queue();
            if (e.getAuthor().getIdLong() == Constants.FLOID) {
                PrivateCommands.execute(e.getMessage());
            }
            PrivateCommand.check(e);
            String msg = e.getMessage().getContentDisplay();
            if (msg.contains("https://") || msg.contains("http://")) {
                Optional<String> urlop = Arrays.stream(msg.split("\n")).filter(s -> s.contains("https://replay.pokemonshowdown.com") || s.contains("http://florixserver.selfhost.eu:228/")).map(s -> s.substring(s.indexOf("http"), s.indexOf(" ", s.indexOf("http") + 1) == -1 ? s.length() : s.indexOf(" ", s.indexOf("http") + 1))).findFirst();
                if (urlop.isPresent()) {
                    String url = urlop.get();
                    analyseReplay(url, null, e.getJDA().getTextChannelById(820359155612254258L), e.getMessage(), null);
                }
            }
        }
    }

    @Override
    public void onGuildVoiceLeave(GuildVoiceLeaveEvent e) {
        if (e.getMember().getId().equals("723829878755164202")) {
            GuildMusicManager manager = getGuildAudioPlayer(e.getGuild());
            manager.scheduler.queue.clear();
            manager.scheduler.nextTrack();
        }
        if (e.getChannelLeft().getMembers().size() == 1 && e.getGuild().getAudioManager().isConnected()) {
            e.getGuild().getAudioManager().closeAudioConnection();
        }
    }

    @Override
    public void onReady(@NotNull ReadyEvent e) {
        try {
            JDA jda = e.getJDA();
            if (jda.getSelfUser().getIdLong() == 723829878755164202L) {
                Draft.init(jda);
                DBManagers.BAN.forAll(rs -> Command.banTimer(jda.getGuildById(rs.getLong("guildid")), Optional.ofNullable(rs.getTimestamp("expires")).map(Timestamp::getTime).orElse(-1L), rs.getLong("userid")));
                DBManagers.MUTE.forAll(rs -> Command.muteTimer(jda.getGuildById(rs.getLong("guildid")), Optional.ofNullable(rs.getTimestamp("expires")).map(Timestamp::getTime).orElse(-1L), rs.getLong("userid")));
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public void onUserUpdateName(UserUpdateNameEvent e) {
        logger.info(e.getOldName() + " -> " + e.getNewName());
        if (e.getUser().getMutualGuilds().stream().map(ISnowflake::getIdLong).anyMatch(l -> l == Constants.ASLID))
            e.getJDA().getTextChannelById("728675253924003870").sendMessage(e.getOldName() + " hat sich auf ganz Discord in " + e.getNewName() + " umbenannt!").queue();
    }

    @Override
    public void onChannelCreate(@NotNull ChannelCreateEvent e) {
        if (e.getChannelType().isGuild()) {
            Guild g = e.getGuild();
            GuildChannel channel = (GuildChannel) e.getChannel();
            if (g.getId().equals("447357526997073930"))
                channel.getPermissionContainer().upsertPermissionOverride(g.getRoleById("761723664273899580")).setDenied(Permission.MESSAGE_SEND).queue();
        }
    }

    @Override
    public void onGuildInviteCreate(GuildInviteCreateEvent e) {
        Guild g = e.getGuild();
        if (g.getIdLong() != Constants.ASLID) return;
        g.retrieveMember(e.getInvite().getInviter()).queue(mem -> {
            if (g.getSelfMember().canInteract(mem)) e.getInvite().delete().queue();
        });
    }

    @Override
    public void onCommandAutoCompleteInteraction(@NotNull CommandAutoCompleteInteractionEvent e) {
        if (e.getCommandType() == net.dv8tion.jda.api.interactions.commands.Command.Type.SLASH) {
            AutoCompleteQuery focusedOption = e.getFocusedOption();
            ArgumentManagerTemplate.Argument arg = byName(e.getName()).getArgumentTemplate().find(focusedOption.getName());
            ArgumentType type = arg.getType();
            if (type.hasAutoComplete()) {
                e.replyChoiceStrings(type.autoCompleteList(focusedOption.getValue())).queue();
            }
        }
    }
}

File: src/main/kotlin/de/tectoast/emolga/bot/EmolgaListener.kt
package de.tectoast.emolga.bot

import de.tectoast.emolga.buttons.ButtonListener
import de.tectoast.emolga.commands.Command
import de.tectoast.emolga.commands.GuildCommandEvent
import de.tectoast.emolga.commands.PrivateCommand
import de.tectoast.emolga.commands.marker
import de.tectoast.emolga.commands.showdown.ReplayCommand
import de.tectoast.emolga.database.exposed.BanDB
import de.tectoast.emolga.database.exposed.MuteDB
import de.tectoast.emolga.modals.ModalListener
import de.tectoast.emolga.selectmenus.MenuListener
import de.tectoast.emolga.utils.Constants
import de.tectoast.emolga.utils.Constants.FLOID
import de.tectoast.emolga.utils.json.db
import de.tectoast.emolga.utils.json.emolga.draft.League
import dev.minn.jda.ktx.events.listener
import dev.minn.jda.ktx.messages.reply_
import mu.KotlinLogging
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.channel.ChannelType
import net.dv8tion.jda.api.entities.emoji.Emoji
import net.dv8tion.jda.api.events.guild.GuildJoinEvent
import net.dv8tion.jda.api.events.guild.invite.GuildInviteCreateEvent
import net.dv8tion.jda.api.events.guild.voice.GuildVoiceUpdateEvent
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.StringSelectInteractionEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.events.role.RoleCreateEvent
import net.dv8tion.jda.api.events.session.ReadyEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.interactions.components.buttons.Button
import org.litote.kmongo.eq

object EmolgaListener : ListenerAdapter() {

    fun registerEvents(jda: JDA) {
        jda.listener<ButtonInteractionEvent> { ButtonListener.check(it) }
        jda.listener<ModalInteractionEvent> { ModalListener.check(it) }
        jda.listener<StringSelectInteractionEvent> { MenuListener.check(it) }
        jda.listener<MessageReceivedEvent> { messageReceived(it) }
        jda.listener<SlashCommandInteractionEvent> { slashCommandInteractionEvent(it) }
        logger.info("important".marker, "Registering ready event...")
        jda.listener<ReadyEvent> { e ->
            logger.info("important".marker, "Ready event received!")
            if (e.jda.selfUser.idLong == 723829878755164202) {
                logger.info("important".marker, "Emolga is now online!")
                db.drafts.find(League::isRunning eq true).toList().forEach {
                    if (it.noAutoStart) return@forEach
                    logger.info("important".marker, "Starting draft ${it.leaguename}...")
                    it.startDraft(null, true, null)
                }
            }
        }
        jda.listener<CommandAutoCompleteInteractionEvent> { e ->
            if (e.commandType == net.dv8tion.jda.api.interactions.commands.Command.Type.SLASH) {
                val focusedOption = e.focusedOption
                val arg = Command.byName(e.name)!!.argumentTemplate.findForAutoComplete(focusedOption.name)!!
                val type = arg.type
                if (type.hasAutoComplete()) {
                    type.autoCompleteList(focusedOption.value, e).run {
                        e.replyChoiceStrings(this ?: emptyList()).queue()
                    }
                }
            }
        }
    }

    private suspend fun slashCommandInteractionEvent(e: SlashCommandInteractionEvent) {
        val command = Command.byName(e.name) ?: return
        val u = e.user
        val tco = e.channel.asGuildMessageChannel()
        val mem = e.member!!
        if (command.disabled) return
        val gid = e.guild!!.idLong
        if (!command.checkBot(e.jda)) return
        val check = command.checkPermissions(gid, mem)
        if (check == Command.PermissionCheck.GUILD_NOT_ALLOWED) return
        if (check == Command.PermissionCheck.PERMISSION_DENIED) {
            e.reply_(Command.NOPERM, ephemeral = true).queue()
            return
        }
        if (mem.idLong != FLOID && mem.idLong != 728202578353193010) {
            if (Command.BOT_DISABLED) {
                e.reply(Command.DISABLED_TEXT).queue()
                return
            }
            if (command.wip) {
                e.reply("Diese Funktion ist derzeit noch in Entwicklung und ist noch nicht einsatzbereit!").queue()
                return
            }
        }
        try {
            GuildCommandEvent(command, e).execute()
        } catch (ex: Command.MissingArgumentException) {
            val arg = ex.argument!!
            e.reply(
                arg.customErrorMessage
                    ?: """Das benötigte Argument `${arg.name}`, was eigentlich ${Command.buildEnumeration(arg.type.getName())} sein müsste, ist nicht vorhanden!
Nähere Informationen über die richtige Syntax für den Command erhältst du unter `e!help ${command.name}`.""".trimIndent()
            ).queue()
            if (u.idLong != FLOID) {
                Command.sendToMe("MissingArgument " + tco.asMention + " Server: " + tco.guild.name)
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
            e.hook.sendMessage(
                """
    Es ist ein Fehler beim Ausführen des Commands aufgetreten!
    Wenn du denkst, dass dies ein interner Fehler beim Bot ist, melde dich bitte bei ${Constants.MYTAG}.
    ${command.getHelp(e.guild)}${if (u.idLong == FLOID) "\nJa Flo, du sollst dich auch bei ihm melden du Kek! :^)" else ""}
    """.trimIndent()
            ).queue()
        }

    }

    override fun onGuildVoiceUpdate(e: GuildVoiceUpdateEvent) {
        e.channelLeft?.let {
            if (it.members.size == 1 && e.guild.audioManager.isConnected) {
                e.guild.audioManager.closeAudioConnection()
            }
            if (e.channelJoined == null && e.member.idLong == e.jda.selfUser.idLong) {
                val manager = Command.getGuildAudioPlayer(e.guild)
                manager.scheduler.queue.clear()
                manager.scheduler.nextTrack()
            }
        }
    }

    override fun onGuildJoin(e: GuildJoinEvent) {
        val g = e.guild
        e.jda.openPrivateChannelById(g.ownerIdLong).flatMap {
            it.sendMessage(WELCOMEMESSAGE.replace("{USERNAME}", "<@${g.ownerIdLong}>").replace("{SERVERNAME}", g.name))
        }.queue()
        e.jda.openPrivateChannelById(FLOID).flatMap {
            it.sendMessage("${g.name} (${g.id})")
                .setActionRow(Button.primary("guildinvite;" + g.id, "Invite").withEmoji(Emoji.fromUnicode("✉️")))
        }.queue()
    }

    override fun onRoleCreate(e: RoleCreateEvent) {
        when (e.guild.idLong) {
            736555250118295622, 447357526997073930, 518008523653775366 -> e.role.manager.revokePermissions(Permission.CREATE_INSTANT_INVITE)
                .queue()
        }
    }

    private suspend fun messageReceived(e: MessageReceivedEvent) {
        when (e.channelType) {
            ChannelType.TEXT -> {
                if (e.isWebhookMessage) return
                if (e.jda.selfUser.idLong == 849569577343385601) Command.check(e)
            }

            ChannelType.PRIVATE -> {
                if (e.author.isBot) return
                if (e.author.idLong != FLOID) e.jda.getTextChannelById(828044461379682314L)
                    ?.sendMessage(e.author.asMention + ": " + e.message.contentDisplay)?.apply {
                        if (e.message.attachments.isNotEmpty()) addContent("\n\n" + e.message.attachments.joinToString("\n") { it.url })
                    }?.queue()
                PrivateCommand.check(e)
                val msg = e.message.contentDisplay
                if (msg.contains("https://") || msg.contains("http://")) {
                    ReplayCommand.regex.find(msg)?.run {
                        val url = groupValues[0]
                        logger.info(url)
                        Command.analyseReplay(
                            url = url,
                            //customReplayChannel = e.jda.getTextChannelById(999779545316069396),
                            resultchannelParam = e.jda.getTextChannelById(820359155612254258)!!, message = e.message
                        )
                    }
                }
            }

            else -> {}
        }
    }

    override fun onReady(e: ReadyEvent) {
        val jda = e.jda
        if (jda.selfUser.idLong == 723829878755164202L) {
            Command.uninitializedCommands.forEach { Command.sendToMe("No Argument Manager Template: $it") }
            BanDB.schedule(jda)
            MuteDB.schedule(jda)
            //Draft(jda.getTextChannelById(837425828245667841)!!, "NDS", null, fromFile = true, isSwitchDraft = true);
        }
    }

    override fun onGuildInviteCreate(e: GuildInviteCreateEvent) {
        val g = e.guild
        when (g.idLong) {
            Constants.G.ASL -> {
                g.retrieveMember(e.invite.inviter!!).queue { mem: Member ->
                    if (g.selfMember.canInteract(mem)) e.invite.delete().queue()
                }
            }
        }
    }

    private val WELCOMEMESSAGE: String = """
            Hallo **{USERNAME}** und vielen Dank, dass du mich auf deinen Server **{SERVERNAME}** geholt hast!
            Vermutlich möchtest du für deinen Server hauptsächlich, dass die Ergebnisse von Showdown Replays in einen Channel geschickt werden.
            Um mich zu konfigurieren gibt es folgende Möglichkeiten:

            **1. Die Ergebnisse sollen in den gleichen Channel geschickt werden:**
            Einfach `/replaychannel add` in den jeweiligen Channel schreiben

            **2. Die Ergebnisse sollen in einen anderen Channel geschickt werden:**
            `/replaychannel add #Ergebnischannel` in den Channel schicken, wo später die Replays reingeschickt werden sollen (Der #Ergebnischannel ist logischerweise der Channel, wo später die Ergebnisse reingeschickt werden sollen)       
            
            Wenn die Channel eingerichtet worden sind, muss man einfach /replay mit dem Replay-Link in einen Replay-Channel schicken und ich erledige den Rest.

            Falls die Ergebnisse in ||Spoilertags|| geschickt werden sollen, schick irgendwo auf dem Server den Command `/spoilertags` rein. Dies gilt dann serverweit.

            Falls du weitere Fragen oder Probleme hast, schreibe ${Constants.MYTAG} eine PN oder komme auf den Support-Server, dessen Link in meinem Profil steht :)
        """.trimIndent()
    private val logger = KotlinLogging.logger {}

}
